text/1001-websocket-protocol.rst

```diff
@@ -0,0 +1,139 @@
+..
+    Status: Draft
+    Type: Feature
+    Created: 2020-02-19
+    RFC PR: `edgedb/rfcs#4 <https://github.com/edgedb/rfcs/pull/4>`_
+
+======================================
+RFC 0001: The WebSocket-based Protocol
+======================================
+
+This RFC describes the design and implementation of the alternate
+protocol to the EdgeDB based on WebSockets for framing and the initial
+handshake.
+
+
+Motivation
+==========
+
+The primary motivation is to allow various browser-based exploration
+and editing tools for the database.
+
+The secondary motivation is to allow browser-based access to the
+application data. Essentially allowing serverless implementation for
+some applications or parts thereof. The reason this is a secondary
+motivation is that it requires full featured access controls (ACLs)
+which we don't have yet.
+
+Two important side effects of implementing this RFC:
+* it might also help implementing the protocol in some languages which
+  we don't have language bindings yet
+* existing tools like load-balancers can be used to access edgedb in
+  some cases
+* it enables using TLS which we don't have for the main protocol yet
+
+
+Specification
+=============
+
+Changes from the Binary Protocol
+--------------------------------
+
+1. The URL to connect to is ``ws://<host>:<port>/ws/<database-name>``
+2. We use single WebSocket subprotocol name ``edgedb-binary``.
+3. Authentication is done at HTTP level (see Authentication_ section)
+4. WebSocket framing is used as a message delimitation. The first
+   byte indicates a message type, followed by message contents.
+5. When WebSocket connection is established, ClientHandshake_ message
+   is sent without ``database`` and ``user`` params.
+6. Authentication messages are never sent from server
+7. Ping/Pong WebSocket messages might be sent and must be responded.
+
+Other than these changes, whole `protocol flow and structure`_
+is the same as using binary protocol.
+
+.. _ClientHandshake: https://edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-client-handshake
+.. _protocol flow and strucutre: https://edgedb.com/docs/internals/protocol/overview
+
+
+Authentication
+==============
+
+TBD
+
+
+Rejected Alternative Ideas
+==========================
+
+Using Subprotocol Name for Version Negotiation
+----------------------------------------------
+
+This means using string like ``edgedb-binary-0.8`` as a subprotocol name
+for the handshake instead of negotiating the version in a separate
+handshake message.
```

Comment:
I think we should use subprotocol name + version negotiation but for the WS tunnel protocol. Essentially "edgedb.1" means that we're using WS tunnel protocol "1". The binary protocol version is negotiated inside the tunnel once it's established.

It's unlikely we need to bump the tunnel protocol version though, but having the ability to do that is crucial for long-term maintainability.

Reply:
Since subprotocol is just a string, we can always add version later. So if it's unlikely that we ever bump version, I see no reason of adding it.

We should name it `edgedb-tunnel`, though. If we will really go with tuneling.

Reply:
> Since subprotocol is just a string, we can always add version later. So if it's unlikely that we ever bump version, I see no reason of adding it.

We maybe need to implement extra WS-specific auth as part of the tunneling proto. @elprans what do you think (IIRC it was your idea)?

> We should name it edgedb-tunnel, though. If we will really go with tuneling.

+1